### PR TITLE
Add support for `prefix` in password policies

### PR DESCRIPTION
**Pull request by maxb**
It is an industry standard practice to generate secrets that include an
identifying fixed prefix, to make it easier to identify secrets that are
mistakenly leaked into logging and version control systems.

This article on the GitHub blog describes the practice, including links
to Slack and Stripe documentation showing those vendors also observe
this practice: https://github.blog/2021-04-05-behind-githubs-new-authentication-token-formats/

Another widely known example is that Amazon AWS access key IDs start
with `AKIA` (or `ASIA` for short-lived credentials).

The Vault Active Directory secrets engine accidentally implements
the practice already, in its deprecated password generation routine, by
starting all passwords for `?@09AZ` - "accidentally", because the stated
reason is to work around password complexity rules on the server side,
but it also accomplishes this goal too.

However the newer "Password Policies" feature in Vault is incapable of
incorporating fixed prefixes on passwords.

This poses a problem for people who would like to migrate away from
deprecated features - since the warnings generated by
`terraform-provider-vault` are irritatingly unsuppressable - but are
unwilling to sacrifice this security feature.

Therefore, this PR adds in capability to specify a fixed prefix to the
new "Password Policies" generation system.

### CODE REVIEW

**Review by cipherboy — APPROVED** [review]
Looks good from a crypto PoV, but I think this requires Ecosystem approval as it affects the password policy used by the database plugin. Not sure if they'd want tests and/or docs for this.

**Comment by maxb** [comment]
I'm open to working on docs/tests if the general design is agreed OK.

e.g. one thing to settle is whether or not I should also implement `suffix` - I have no intention of using it myself, nor really expect anyone else to need to, but it is needed for full compatibility with the old way of doing things - maybe I should add it in as it would only be a few lines of code more, just in case?

File: helper/random/string_generator.go
Comment (cipherboy) [inline helper/random/string_generator.go:72]: I know we clarified it below, but it might be nice to state that this is the length of the _random_ portion of the password.

File: helper/random/string_generator.go
Comment (austingebauer) [inline helper/random/string_generator.go:78]: I'm wondering if we should restrict the prefix to a fixed set of characters? The reason I ask is because I could see it being used in SQL injection. This is [an example](https://github.com/hashicorp/vault/blob/main/plugins/database/mysql/mysql.go#L24) of where the password is currently string replaced. This doesn't use prepared statements with parameters today (though I think it should).

I understand the likelihood of this happening is low. The password policy and database secrets APIs are authenticated and often configured by a privileged Vault user. In any case, I think we should guard against this with sufficient input validation if possible. Let me know your thoughts.

**Review by austingebauer — COMMENTED** [review]
Thanks for this, @maxb! I'd say we can hold off on adding a suffix. I have one question/concern. Otherwise, it's looking good.

File: helper/random/string_generator.go
Comment (maxb) [reply helper/random/string_generator.go:78]: That is an interesting thought.

However, the password creation policy system is not specific to databases, nor do all databases use SQL, nor are all SQL dialects all that similar, so it could be tricky to come up with a list of characters that meet all use-cases, whilst avoiding user surprise. (e.g. a hypothetical user reaction: "I can't use this character in my LDAP password prefix ... to prevent SQL injection? LDAP doesn't have anything to do with SQL! What's going on?!")

Additionally, the only way I can see this being used for SQL injection, would be if an adversary had already gained privileged access to modify Vault configuration - at which point, there's a fair chance they can just directly modify the database secrets engine configuration too.

My inclination, therefore, is to not attempt to combat SQL injection within the general password creation policy subsystem.

Comment (austingebauer) [reply helper/random/string_generator.go:78]: > However, the password creation policy system is not specific to databases, nor do all databases use SQL, nor are all SQL dialects all that similar, so it could be tricky to come up with a list of characters that meet all use-cases, whilst avoiding user surprise. (e.g. a hypothetical user reaction: "I can't use this character in my LDAP password prefix ... to prevent SQL injection? LDAP doesn't have anything to do with SQL! What's going on?!")

I definitely hear you. Restricting the password policy prefix isn't the right way to solve the underlying issue. I'm wondering if there is something pragmatic we can do to support some prefixes while we address it. Starting with a limited character set still allows us to expand (for when that LDAP user asks :)) in the future. This could allow us to deliver _some_ value today while being able to expand the character set later on.

Do you think that's a reasonable compromise in the short term?

Comment (maxb) [reply helper/random/string_generator.go:78]: Apologies for going silent on this conversation for many months - I have parted ways with the employer who had the original motivating use-case, so I no longer personally expect to make use of this feature any time soon.

As to the design issue being discussed ... I still feel that any character set restriction enforced at this point in the code, would provide insufficient practical benefit, for substantial downsides in extra complexity, documentation needed, and potential user confusion.

* I don't want to put a restriction on the whole password policy subsystem, which is used by many secrets engines, just for the database secrets engine.
* Especially when correctly defining the intersection of safe character sets for every current and future database plugin, is going to be an imperfect compromise anyway.
* And the entire stack of compromises is only meaningful at all, in defending against an adversary who has gained permission to write to `sys/policies/password/*`, but not `sys/policies/acl/*` or `database/config/*` - which, whilst admittedly a possible configuration of Vault, seems very niche indeed.

With those doubts, on top of no longer having a current personal motivation to progress this particular feature, I think it's best if I just mark this PR as draft for now - and, if you would prefer instead to close it, I'm OK with that.

**Comment by maxb** [comment]
I'm going to close this PR for now.

As per the longer comment a couple of months ago in the discussion thread above, I no longer expect to make use of this feature personally, and we did not reach a consensus on what further work is or is not needed to make this mergeable.

**Pull request closed without merging**
